kuka_rsi_hw_interface/src/rsi_command_handler.cpp

```diff
@@ -0,0 +1,325 @@
+        throw std::logic_error(err_buf_);
+      }
+      buffer_it++;
+      if (!element.CastParam(node_param, buffer_it)) {
+        auto i = std::sprintf(err_buf_, "Could not cast the ");
+        std::snprintf(err_buf_ + i, node_param.length_ + 1, "%s", node_param.data_ptr_);
+        i += node_param.length_;
+        std::sprintf(
+          err_buf_ + i, " param into the %s elements parameter list.",
+          element.GetName().c_str());
+        throw std::logic_error(err_buf_);
+      }
+      if (*buffer_it != '"') {
+        auto i = sprintf(err_buf_, "In \"");
+        std::snprintf(err_buf_ + i, node_param.length_ + 1, "%s", node_param.data_ptr_);
+        i += node_param.length_;
+        std::sprintf(err_buf_ + i, "\" param syntax error found");
+        throw std::logic_error(err_buf_);
+      }
+      buffer_it++;
+      numOfParam++;
+    }
+  }
+
+  // Fast forward to close bracket
+  for (; (size_t)(buffer_it - buffer) < buffer_size && *buffer_it != '>'; buffer_it++) {
+  }
+  if ((size_t)(buffer_it - buffer) >= buffer_size) {
+    throw std::range_error("Out of the buffers range");
+  }
+  buffer_it++;
+
+  // Could not find all parameter, or found too much
+  if (isBaseLessNode && numOfParam != element.GetParams().size()) {
+    std::sprintf(
+      err_buf_,
+      "%lu parameter found. It does not match with %s elements parameter list size.", numOfParam,
+      element.GetName().c_str());
+    throw std::logic_error(err_buf_);
+  }
+
+  // Node should have childs, but did not found any
+  if (isBaseLessNode && element.GetChilds().size() > 0) {
+    std::sprintf(
+      err_buf_, "%s node should have chields but did not found any",
+      element.GetName().c_str());
+    throw std::logic_error(err_buf_);
+  }
+  // fast forward if
+  for (; *buffer_it == ' '; buffer_it++) {
+  }
+  if ((size_t)(buffer_it - buffer) >= buffer_size) {
+    throw std::range_error("Out of the buffers range");
+  }
+
+  if (!isBaseLessNode) {
+    if (*buffer_it != '<') {
+      // Node base is data
+      if (!element.CastParam(node_name, buffer_it)) {
+        auto i = std::sprintf(err_buf_, "Could not cast the ");
+        std::snprintf(err_buf_ + i, node_name.length_ + 1, "%s", node_name.data_ptr_);
+        i += node_name.length_;
+        std::sprintf(
+          err_buf_ + i, " parameter into the %s elements parameter list",
+          element.GetName().c_str());
+        throw std::logic_error(err_buf_);
+      }
+      if (*buffer_it != '<') {
+        auto i = sprintf(err_buf_, "In \"");
+        std::snprintf(err_buf_ + i, node_name.length_ + 1, "%s", node_name.data_ptr_);
+        i += node_name.length_;
+        std::sprintf(err_buf_ + i, "\" param syntax error found");
+        throw std::logic_error(err_buf_);
+      }
+    } else {
+      // node base has childs
+      for (auto && child : element.GetChilds()) {
+        decodeNode(child, buffer, buffer_it, buffer_size);
```

Comment:
This looks a bit dangerous. This is a pretty long function with a recursive call in the middle. So I assume that you want to do a depth first search here, pass the iterator around, process the current element and then return to the parent.

1) How do you ensure that the buffer won't overflow?

2) Are exceptions thrown here supposed to be caught or is aborting the desired behavior?

3) Can you provide any clear exit conditions at the beginning for the recursive function? Or is there nothing we can do here?

4) Also, can you somehow make the recursive function happen at the end?

4) I assume this will be called multiple times on the root node. Since we are so real-time we're not even allocating dynamically because it's evil check out https://en.wikipedia.org/wiki/Memoization for huge potential performance improvements. For example, this is also how you ensure how a recursive Fibonacci function doesn't hang for bigger inputs.

You can also use a queue and store the nodes here to avoid recursive calls, I'm fine with this solution but you should be able to reason about the correctness of it.

Reply:
resolved